### PR TITLE
More Strongly-Typed Rust Bindings

### PR DESCRIPTION
**Pull request by kazimuth**
After reading #2523 I was messing with the newly-merged Rust API and noticed some some changes that might be helpful.
The main change this PR makes is to start adding some "typed" rust APIs, which look like normal rust functions / structs, and call tvm PackedFuncs under the hood. The rust graph runtime example is now much simpler:

```rust
    let image: ndarray::Array = /* ... load image in Rust ... */;

    let input = tvm_frontend::NDArray::from_rust_ndarray(
        &image,
        TVMContext::cpu(0),
        TVMType::from("float32")
    )?;

    let mut module = GraphModule::from_paths(
        concat!(env!("CARGO_MANIFEST_DIR"), "/deploy_graph.json"),
        concat!(env!("CARGO_MANIFEST_DIR"), "/deploy_lib.so"),
        concat!(env!("CARGO_MANIFEST_DIR"), "/deploy_param.params"),
        ctx
    )?;

    let mut result: Vec<tvm_frontend::NDArray> = module.apply(&[&input]);
```

Of course, all the old PackedFunc functionality is still there. This just makes well-defined APIs a little easier to use.

This approach could be used for binding the non-runtime APIs as well, I think. (Which I want, because I kinda wanna make a tvm-backed rust deep learning framework at some point... that's the far future though ;))

I've also fixed up a few other random things I've noticed:
- I think the old `Function::get` implementation would break for freshly registered globals, so I made it more dynamic
- I allowed `kNDArrayContainer`s to be returned from PackedFuncs without blowing up... probably. The example works, anyway. I don't entirely understand what's going on with the stuff in `common::value` though.

TODO:
- [ ] more docs
- [ ] more tests
- [ ] run rustfmt

Open questions:
- What other APIs would be good to wrap with types?

CC @ehsanmok , hope it's okay that i'm poking around in your code :)

### CODE REVIEW

**Comment by ehsanmok** [comment]
Thank you for testing out and making nice suggestions! any suggestions is more than welcome of course :)
I agree! `GraphModule` and `GraphRuntime` are more typed and ergonomic and they are part of the next plan, glad you've mentioned them. For example, you can also see the java equivalent in [contrib](https://github.com/dmlc/tvm/tree/master/jvm/core/src/main/java/ml/dmlc/tvm/contrib). 

Fyi @nhynes has been working on some more ergonomic and enhancements for `common` and few other things as part of his RFC that you linked to.

> What other APIs would be good to wrap with types?

So far I've kept the compatibility with other frontends (python, java and golang) in mind, but that doesn't mean we can't add more. Off the top of my head, when [#2498](https://github.com/dmlc/tvm/pull/2498) is merged it could open up some good opportunities in Rust.

**Comment by nhynes** [comment]
Hey @kazimuth thanks a lot for the PR! This is really cool! It looks like you've made a lot of overlapping changes with my [branch which unifies `frontend` and `runtime`](https://github.com/nhynes/tvm/pull/4/files). Would you mind if I asked you to postpone further work until I get that fully tested and merged? I'll prioritize it for this week. I can't stop you from contributing, but it will save us both a lot of work if we collaborated on the unified `tvm` crate (e.g., I [already fixed the globals dict](https://github.com/nhynes/tvm/pull/4/files#diff-8e31becb859121f6397ff2eca51d9d41R24)). If it makes you more willing to wait, I'll apply the semantics of your commit to the updated branch! What do you think?

File: rust/frontend/README.md [hunk 1]
```diff
@@ -44,50 +44,31 @@ Now, we need to input the artifacts to create and run the *Graph Runtime* to det
...
-    let runtime_create_fn_ret = call_packed!(
-        runtime_create_fn,
-        &graph,
-        &lib,
-        &ctx.device_type,
-        &ctx.device_id
+    let image: ndarray::Array = /* ... load image in Rust ... */;
+
+    // convert to a tvm NDArray
+    // could be on a GPU, FPGA, ...
```
Comment (nhynes) [inline rust/frontend/README.md]: mind if I ask which exactly you're trying to use? ;)

**Comment by nhynes** [comment]
you'll want to `cargo install rustfmt-nightly --version 1.0.1 --force`. I think that your local rustfmt might not be running :)

**Comment by nhynes** [comment]
also, rustfmt-on-save using [rust.vim](https://github.com/rust-lang/rust.vim) is 👌

**Comment by kazimuth** [comment]
Thanks for the advice @nhynes, I'll delay until you're done with that change -- didn't see your PR :) and no rush, I was just poking at things and figured I could chip in.

File: rust/frontend/src/function.rs [hunk 2]
```diff
@@ -18,21 +18,8 @@ use std::{
...
-        ));
-        let names_list = unsafe { slice::from_raw_parts(out_array, out_size as usize) };
-        Mutex::new(
-            names_list
-                .into_iter()
-                .map(|&p| (unsafe { CStr::from_ptr(p).to_str().unwrap() }, None))
-                .collect(),
-        )
+    static ref GLOBAL_FUNCTIONS: Mutex<HashMap<String, Function>> = {
+        Mutex::new(HashMap::new())
```
[the remark this answers is not in the document]
Comment (kazimuth) [reply rust/frontend/src/function.rs]: I believe `TVMFuncRegisterGlobal` is meant for adding new global functions at runtime, which means that the set of names may change?

File: rust/frontend/README.md [hunk 3]
```diff
@@ -44,50 +44,31 @@ Now, we need to input the artifacts to create and run the *Graph Runtime* to det
...
-    // flatten the output as Vec<f32>
-    let output = output.to_vec::<f32>()?;
+    // load the modules
+    // these modules can also be provided as raw blobs of bytes instead of file paths,
+    // if you want to store the data in a different way, like embedded in the executable,
+    // or on the network.
+    let mut module = GraphModule::from_paths(
+        concat!(env!("CARGO_MANIFEST_DIR"), "/deploy_graph.json"),
+        concat!(env!("CARGO_MANIFEST_DIR"), "/deploy_lib.so"),
+        concat!(env!("CARGO_MANIFEST_DIR"), "/deploy_param.params"),
```
[the remark this answers is not in the document]
Comment (kazimuth) [reply rust/frontend/README.md]: Makes sense, I was just aiming to minimize lines of example code.

I wasn't sure about providing from_paths at all because it won't work in e.g. wasm. Might make sense to put behind a `#[cfg]`.

File: rust/frontend/src/function.rs [hunk 4]
```diff
@@ -66,25 +53,33 @@ impl Function {
...
+            });
+        }
+        let c_name = CString::new(name)?;
+        let mut handle = ptr::null_mut() as ts::TVMFunctionHandle;
+        check_call!(ts::TVMFuncGetGlobal(
+            c_name.as_ptr() as *const c_char,
+            &mut handle as *mut _
+        ));
+        if handle.is_null() {
+            bail!(ErrorKind::FunctionNotFound(name.to_string()));
```
[the remark this answers is not in the document]
Comment (kazimuth) [reply rust/frontend/src/function.rs]: Ooh, nice

File: rust/frontend/src/graph_runtime.rs [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+    pub fn new(graph: &str, lib: &Module, ctx: TVMContext) -> Result<Self> {
+        let module = packed::create(graph, lib, &ctx.device_type, &ctx.device_id)?;
+
+        let set_input_ = module.get_function("set_input", false).expect("no set_input, invalid GraphModule");
+        let get_input_ = module.get_function("get_input", false).expect("no get_input, invalid GraphModule");
+        let get_output_ = module.get_function("get_output", false).expect("no get_output, invalid GraphModule");
+        let get_num_outputs_ = module.get_function("get_num_outputs", false).expect("no get_num_outputs, invalid GraphModule");
+        let run_ = module.get_function("run", false).expect("no run_, invalid GraphModule");
+        let load_params_ = module.get_function("load_params", false).expect("no load_params_, invalid GraphModule");
```
[the remark this answers is not in the document]
Comment (kazimuth) [reply rust/frontend/src/graph_runtime.rs]: My understanding was that `Module` is a type wrapping a dynamic library that could have any number of named functions, e.g. one created with `tvm.contrib.cc`; whereas `GraphModule` is a particular API that some `Module`s expose. So it made sense to make GraphModule a simple typed Module wrapper.

[the remark this answers is not in the document]
Comment (nhynes) [reply rust/frontend/src/graph_runtime.rs]: kind of. there is really no graph api inside of the module: it's purely defined by the graph.json. For instance, a module with `dense1`, `batch_norm`, and `softmax` could create a graph with {`dense1`, `softmax} just as easily as with all three. Basically the Module is a repo of functions and the graph.json determines how those are actually called.

File: rust/frontend/src/function.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (nhynes) [reply rust/frontend/src/function.rs]: Yes, they could change, but functions can also be removed which makes caching _at all_ ill advised. The de-facto semantics of TVM globals is static registration. Always re-wrapping pointers might be an option but I'm not sure if the performance cost would be worth the (unused) additional flexibility.

On the bright side, part of the reason for the rust runtime has been to remove these pesky globals. As we further expand the rust functionality, we won't have to think as much about C++ interop.

Once I finish up unifying the tvm crate and merging in your PR, a good next area of focus might be on creating device apis to enable use cases like GPUs (let's start with AMD because you seem to have a real use case) and VTA.

**Comment by kazimuth** [comment]
Hey @nhynes, any update on your PR? No rush, just curious

**Comment by nhynes** [comment]
Hey @kazimuth super sorry for the long delay. I was planning on taking a TVM day tomorrow and Sunday. I don't want to block your contribution since it contains a number of useful improvements, so if I don't have the merge done by Monday, I'll merge this and rebase. Thanks!

**Comment by ehsanmok** [comment]
@kazimuth Would you want to still work on adding `GraphModule` ? it's a nice feature to have.

**Comment by nhynes** [comment]
I think that what we actually want is a `GraphExecutor` for the frontend. We've already established that there's no such thing as a "GraphModule" _per se_, but rather a module that  exports functions that can be assembled into an executor when provided along with a graph.json.

@kazimuth if you're still interested in working on improving the frontend graph runtime, we should discuss the design of this component. It'd be great if we could start using conditional compilation to make  a single `runtime` that uses either bindings or Rust.

**Comment by ehsanmok** [comment]
@nhynes In case of frontend, I think `GraphModule` would still be useful and we should be compatible with other frontends such as [java which has it](https://github.com/dmlc/tvm/blob/master/jvm/core/src/main/java/ml/dmlc/tvm/contrib/GraphModule.java). At least it makes writing the application example more succinct.

**Comment by nhynes** [comment]
I'm not disagreeing with the utility, but the name leaves to be desired since it's decisively _not_ a module in the TVM sense. Regardless of nomenclature, we only need one thing that provides the graph runtime/module/executor abstraction.

**Comment by kazimuth** [comment]
If you can give me a broad-strokes design (like some type names + general APIs) I'd be happy to implement it.

> It'd be great if we could start using conditional compilation to make a single runtime that uses either bindings or Rust.

I'm not sure I understand what this means, could you elaborate?

**Comment by nhynes** [comment]
So from a high level I mean something like
```rust
impl GraphExecule { // TODO: name
	fn set_input(&mut self, input: Tensor) {
        
		#[cfg(feature = "bindings")]
		{
            module.get_function("set_input", ...);
		}
		#[cfg(not(feature = "bindings")]
        {
            // pure rust impl
        }
	}

	// ...
}
```

**Comment by kazimuth** [comment]
ah, i guess i'm still a little confused about the distinction between bindings and pure-rust; i thought that distinction had been eliminated. Is pure-rust just for, like, the wasm and sgx runtimes, or is it going to have a bigger application area?

Side note, I think I'm going to close this PR and start a new one for the new changes.

**Comment by nhynes** [comment]
> distinction had been eliminated

It's a constant WIP. The two crates now share common types, but I suspect that there's a decent amount of duplicated logic. I'd also like for users to be able to do `tvm::GraphModule` instead of `tvm::frontime::GraphSomething`. The end goal, of course, would be load the dylib using Rust [`libloading`](https://crates.io/crates/libloading) and never call C++.

Actually, if you're feeling ambitious, you could actually go and implement a shared object module in `tvm_runtime` and update the `GraphExecutor` UI. I guess it also depends on whether you actually want to use the C++ impl.

Also, I remember you saying that you wanted an easy-to-use Rust ML library. I had a neat idea that you might like: an option to build the entire GraphExecutor at compile-time from a pre-defined TVM module and graph json. Like there'll be a macro that generates all of the `extern fn`s, closures calling them, and allocations statically. These can only be added to the Rust runtime, though.

**Comment by kazimuth** [comment]
> The end goal, of course, would be load the dylib using Rust libloading and never call C++.

Is it possible to call a PackedFunc from a deployed library without using C++? Don't you need to go through the C runtime API to call any PackedFunc? Or is that actually implemented somewhere in the rust codebase?

> Also, I remember you saying that you wanted an easy-to-use Rust ML library. I had a neat idea that you might like: an option to build the entire GraphExecutor at compile-time from a pre-defined TVM module and graph json. Like there'll be a macro that generates all of the extern fns, closures calling them, and allocations statically. These can only be added to the Rust runtime, though.

Oh that is an interesting idea. It would be easier if GraphExecutor was already fully-rust too.

Question: Is the current deploy_graph.json / deploy_lib.so format stable, or is it going to be replaced by a new format to support Relay? I was under the impression it was based on the old NNVM graphs, not sure it's worth the effort to port if it's going to be immediately deprecated.

**Comment by nhynes** [comment]
> Is it possible to call a PackedFunc from a deployed library without using C++? 

Yes, if there were a way to load the functions from a dynamic library. The standalone runtime already does this for statically linked module libraries.

> GraphExecutor was already fully-rust too.

Well, strictly speaking, the GraphExecutor only exist in the standalone (Rust) runtime :)

> Is the current deploy_graph.json / deploy_lib.so format stable

Yes, Relay produces the same output as NNVM. At some point, we might need to add support for the more interesting control flow operators, though. That's still WIP on the C++ side AFAIK.

**Comment by kazimuth** [comment]
Closing this because it's bitrotted, will re-start this work later

**Comment by ehsanmok** [comment]
Hi @kazimuth , just pinging to see if you're still interested and would like to propose a deadline for it?

**Comment by kazimuth** [comment]
I am but my workload's picked up a bit, might be a few weeks before i get back to this :)
